### PR TITLE
Sliding window changes

### PR DESCRIPTION
**Pull request by mibur1**
Hi @mtorabi59 

I implemented the changes as discussed in #31.

The changes are:
- Removed extending the window which has two effects:
    - Bugfix regarding the zero-padding for rectangular SW calculation
    - Limits the tapered window to contain no samples outside the window width
- Added a `window_std` parameter to Sliding Window, Discrete HMM, and Sliding Window Clustering
    - Default is `None`, which results to the original std of  `3 * W / 22`
- Disabled the convolution and instead directly use the Gaussian
    - In my opinion this is unnecessary now, as the window size is a hard limit for the window
- Vectorized Pearson correlation calculation for improved performance

Feel free to modify this to your liking. Also I did not write any tests/documentation, please let me know if I should do so :)

### CODE REVIEW

**Comment by mtorabi59** [comment]
Thanks @mibur1 ! Overall it looks good! Thank you for making the changes. There a few comments below:

File: pydfc/dfc_methods/sliding_window.py [hunk 1]
```diff
@@ -95,39 +96,28 @@ def calc_MI(self, X, Y):
...
-            mean = np.mean(time_series, axis=1, keepdims=True)
-            std = np.std(time_series, axis=1, keepdims=True)
-            time_series_standardized = np.where(std != 0, (time_series - mean) / std, 0)
-            model = GraphicalLasso(alpha=self.graphical_lasso_alpha_)
-            model.fit(time_series_standardized.T)
-            # the covariance matrix will equal the correlation matrix
+        # Graphical Lasso
+        if self.params['sw_method']=='GraphLasso':
+            model = GraphicalLassoCV()
+            model.fit(time_series.T)
```
Comment (mtorabi59) [inline pydfc/dfc_methods/sliding_window.py:102]: This part shouldn't change. It's probably because you didn't merge the final version before making the changes.

File: pydfc/dfc_methods/sliding_window.py [hunk 2]
```diff
@@ -95,39 +96,28 @@ def calc_MI(self, X, Y):
...
-                            C[j, i] = 0
-                        else:
-                            C[j, i] = np.corrcoef(X, Y)[0, 1]
-
-                    C[i, j] = C[j, i]
+                    C[j, i] = self.calc_MI(X, Y)
+        # Pearson correlation
+        else:
+            C = np.corrcoef(time_series)
+            C[np.isnan(C)] = 0
```
Comment (mtorabi59) [inline pydfc/dfc_methods/sliding_window.py:117]:
```suggestion
# make the diagonal elements 1 (for nan values on the diagonal)
C[np.diag_indices_from(C)] = 1
```

File: pydfc/dfc_methods/sliding_window.py [hunk 3]
```diff
@@ -141,36 +131,25 @@ def dFC(self, time_series, W=None, n_overlap=None, tapered_window=False):
...
-            FCSs.append(
-                self.FC(
-                    np.multiply(time_series, window)[
-                        :, max(int(l - W / 2), 0) : min(int(l + 3 * W / 2), L)
-                    ]
-                )
-            )
+                std = window_std if window_std is not None else 3 * W / 22
+                window_taper = signal.windows.gaussian(W, std=std)
+                #window = signal.convolve(window, window_taper, mode="same") / sum(window_taper)
```
Comment (mtorabi59) [inline pydfc/dfc_methods/sliding_window.py:146]: Please note that the result of the this is different from what is described in Allen et al. paper. A tapered window is a rectangle convolved with a gaussian not a gaussian.

File: pydfc/dfc_methods/sliding_window.py [hunk 4]
```diff
@@ -141,36 +131,25 @@ def dFC(self, time_series, W=None, n_overlap=None, tapered_window=False):
...
+                window[l:l+W] = window_taper
+
```
Comment (mtorabi59) [inline pydfc/dfc_methods/sliding_window.py:148]:
```suggestion
            # Taper the window
            if tapered_window:
                window = signal.convolve(window, window_taper, mode="same") / sum(window_taper)
```

File: pydfc/dfc_methods/sliding_window.py [hunk 5]
```diff
@@ -141,36 +131,25 @@ def dFC(self, time_series, W=None, n_overlap=None, tapered_window=False):
             model.fit(time_series.T)
             self.graphical_lasso_alpha_ = model.alpha_
```
Comment (mtorabi59) [inline pydfc/dfc_methods/sliding_window.py:133]:
```suggestion
        if tapered_window:
            std = window_std if window_std is not None else 3 * W / 22
            window_taper = signal.windows.gaussian(W, std=std)
```

**Comment by mtorabi59** [comment]
@mibur1 lmk if you have any Qs! 
Also, tests are not necessary, but if you have some ideas for testing the sliding window, it'd be great to have them!

**Comment by mibur1** [comment]
@mtorabi59 Thanks for the quick review and comments, I hope I addressed all of them. I will think about tests for the sliding window and add some if I have an idea.

**Comment by mtorabi59** [comment]
@mibur1 Thanks, everything looks good! There might only be some formatting changes by black or flake8.

**Comment by mtorabi59** [comment]
@mibur1  If you're familiar with pre-commit, it would be great to run it locally and make the required formatting changes, ow I'll fix them!

**Comment by mibur1** [comment]
Thanks! I'll try to do this as well as the unit tests next week. But please feel free to do it yourself if you would like to have it done earlier.

**Comment by mtorabi59** [comment]
@mibur1 Thanks !

**Comment by mtorabi59** [comment]
@mibur1 Since the remaining changes are very small and you have done most of the work, I'm just gonna make the final adjustments myself.
Thanks again for your contribution!

**Comment by mibur1** [comment]
Sorry, I forgot about that. Thanks a lot for handling it!
